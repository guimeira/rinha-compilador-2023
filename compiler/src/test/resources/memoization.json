{"name":"memoization.rinha","expression":{"kind":"Let","name":{"text":"pure","location":{"start":4,"end":8,"filename":"memoization.rinha"}},"value":{"kind":"Function","parameters":[{"text":"x","location":{"start":14,"end":15,"filename":"memoization.rinha"}}],"value":{"kind":"Binary","lhs":{"kind":"Var","text":"x","location":{"start":24,"end":25,"filename":"memoization.rinha"}},"op":"Add","rhs":{"kind":"Int","value":2,"location":{"start":28,"end":29,"filename":"memoization.rinha"}},"location":{"start":24,"end":29,"filename":"memoization.rinha"}},"location":{"start":11,"end":31,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"impure","location":{"start":37,"end":43,"filename":"memoization.rinha"}},"value":{"kind":"Function","parameters":[{"text":"x","location":{"start":49,"end":50,"filename":"memoization.rinha"}}],"value":{"kind":"Let","name":{"text":"_","location":{"start":63,"end":64,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Var","text":"x","location":{"start":73,"end":74,"filename":"memoization.rinha"}},"location":{"start":67,"end":75,"filename":"memoization.rinha"}},"next":{"kind":"Binary","lhs":{"kind":"Var","text":"x","location":{"start":79,"end":80,"filename":"memoization.rinha"}},"op":"Add","rhs":{"kind":"Int","value":2,"location":{"start":83,"end":84,"filename":"memoization.rinha"}},"location":{"start":79,"end":84,"filename":"memoization.rinha"}},"location":{"start":59,"end":84,"filename":"memoization.rinha"}},"location":{"start":46,"end":86,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"impure_indirect","location":{"start":92,"end":107,"filename":"memoization.rinha"}},"value":{"kind":"Function","parameters":[{"text":"x","location":{"start":113,"end":114,"filename":"memoization.rinha"}}],"value":{"kind":"Call","callee":{"kind":"Var","text":"impure","location":{"start":123,"end":129,"filename":"memoization.rinha"}},"arguments":[{"kind":"Var","text":"x","location":{"start":130,"end":131,"filename":"memoization.rinha"}}],"location":{"start":123,"end":132,"filename":"memoization.rinha"}},"location":{"start":110,"end":134,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"maybe_impure","location":{"start":140,"end":152,"filename":"memoization.rinha"}},"value":{"kind":"Function","parameters":[{"text":"closure","location":{"start":158,"end":165,"filename":"memoization.rinha"}},{"text":"x","location":{"start":167,"end":168,"filename":"memoization.rinha"}}],"value":{"kind":"Call","callee":{"kind":"Var","text":"closure","location":{"start":177,"end":184,"filename":"memoization.rinha"}},"arguments":[{"kind":"Var","text":"x","location":{"start":185,"end":186,"filename":"memoization.rinha"}}],"location":{"start":177,"end":187,"filename":"memoization.rinha"}},"location":{"start":155,"end":189,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":195,"end":196,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"pure","location":{"start":205,"end":209,"filename":"memoization.rinha"}},"arguments":[{"kind":"Int","value":2,"location":{"start":210,"end":211,"filename":"memoization.rinha"}}],"location":{"start":205,"end":212,"filename":"memoization.rinha"}},"location":{"start":199,"end":213,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":219,"end":220,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"pure","location":{"start":229,"end":233,"filename":"memoization.rinha"}},"arguments":[{"kind":"Int","value":2,"location":{"start":234,"end":235,"filename":"memoization.rinha"}}],"location":{"start":229,"end":236,"filename":"memoization.rinha"}},"location":{"start":223,"end":237,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":244,"end":245,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"impure","location":{"start":254,"end":260,"filename":"memoization.rinha"}},"arguments":[{"kind":"Int","value":2,"location":{"start":261,"end":262,"filename":"memoization.rinha"}}],"location":{"start":254,"end":263,"filename":"memoization.rinha"}},"location":{"start":248,"end":264,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":270,"end":271,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"impure","location":{"start":280,"end":286,"filename":"memoization.rinha"}},"arguments":[{"kind":"Int","value":2,"location":{"start":287,"end":288,"filename":"memoization.rinha"}}],"location":{"start":280,"end":289,"filename":"memoization.rinha"}},"location":{"start":274,"end":290,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":297,"end":298,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"impure_indirect","location":{"start":307,"end":322,"filename":"memoization.rinha"}},"arguments":[{"kind":"Int","value":2,"location":{"start":323,"end":324,"filename":"memoization.rinha"}}],"location":{"start":307,"end":325,"filename":"memoization.rinha"}},"location":{"start":301,"end":326,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":332,"end":333,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"impure_indirect","location":{"start":342,"end":357,"filename":"memoization.rinha"}},"arguments":[{"kind":"Int","value":2,"location":{"start":358,"end":359,"filename":"memoization.rinha"}}],"location":{"start":342,"end":360,"filename":"memoization.rinha"}},"location":{"start":336,"end":361,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":368,"end":369,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"maybe_impure","location":{"start":378,"end":390,"filename":"memoization.rinha"}},"arguments":[{"kind":"Var","text":"pure","location":{"start":391,"end":395,"filename":"memoization.rinha"}},{"kind":"Int","value":20,"location":{"start":396,"end":398,"filename":"memoization.rinha"}}],"location":{"start":378,"end":399,"filename":"memoization.rinha"}},"location":{"start":372,"end":400,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":406,"end":407,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"maybe_impure","location":{"start":416,"end":428,"filename":"memoization.rinha"}},"arguments":[{"kind":"Var","text":"pure","location":{"start":429,"end":433,"filename":"memoization.rinha"}},{"kind":"Int","value":20,"location":{"start":434,"end":436,"filename":"memoization.rinha"}}],"location":{"start":416,"end":437,"filename":"memoization.rinha"}},"location":{"start":410,"end":438,"filename":"memoization.rinha"}},"next":{"kind":"Let","name":{"text":"_","location":{"start":445,"end":446,"filename":"memoization.rinha"}},"value":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"maybe_impure","location":{"start":455,"end":467,"filename":"memoization.rinha"}},"arguments":[{"kind":"Var","text":"impure","location":{"start":468,"end":474,"filename":"memoization.rinha"}},{"kind":"Int","value":20,"location":{"start":475,"end":477,"filename":"memoization.rinha"}}],"location":{"start":455,"end":478,"filename":"memoization.rinha"}},"location":{"start":449,"end":479,"filename":"memoization.rinha"}},"next":{"kind":"Print","value":{"kind":"Call","callee":{"kind":"Var","text":"maybe_impure","location":{"start":487,"end":499,"filename":"memoization.rinha"}},"arguments":[{"kind":"Var","text":"impure","location":{"start":500,"end":506,"filename":"memoization.rinha"}},{"kind":"Int","value":20,"location":{"start":507,"end":509,"filename":"memoization.rinha"}}],"location":{"start":487,"end":510,"filename":"memoization.rinha"}},"location":{"start":481,"end":511,"filename":"memoization.rinha"}},"location":{"start":441,"end":511,"filename":"memoization.rinha"}},"location":{"start":402,"end":511,"filename":"memoization.rinha"}},"location":{"start":364,"end":511,"filename":"memoization.rinha"}},"location":{"start":328,"end":511,"filename":"memoization.rinha"}},"location":{"start":293,"end":511,"filename":"memoization.rinha"}},"location":{"start":266,"end":511,"filename":"memoization.rinha"}},"location":{"start":240,"end":511,"filename":"memoization.rinha"}},"location":{"start":215,"end":511,"filename":"memoization.rinha"}},"location":{"start":191,"end":511,"filename":"memoization.rinha"}},"location":{"start":136,"end":511,"filename":"memoization.rinha"}},"location":{"start":88,"end":511,"filename":"memoization.rinha"}},"location":{"start":33,"end":511,"filename":"memoization.rinha"}},"location":{"start":0,"end":511,"filename":"memoization.rinha"}},"location":{"start":0,"end":511,"filename":"memoization.rinha"}}
