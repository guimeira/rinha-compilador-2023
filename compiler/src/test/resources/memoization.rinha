let pure = fn(x) => {
  x + 2
};
let impure = fn(x) => {
  let _ = print(x);
  x + 2
};
let impure_indirect = fn(x) => {
  impure(x)
};
let maybe_impure = fn(closure, x) => {
  closure(x)
};
let _ = print(pure(2));
let _ = print(pure(2));

let _ = print(impure(2));
let _ = print(impure(2));

let _ = print(impure_indirect(2));
let _ = print(impure_indirect(2));

let _ = print(maybe_impure(pure,20));
let _ = print(maybe_impure(pure,20));

let _ = print(maybe_impure(impure,20));
print(maybe_impure(impure,20))