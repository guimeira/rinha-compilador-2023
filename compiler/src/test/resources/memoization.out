4
4
2
4
2
4
2
4
2
4
22
22
20
22
20
22